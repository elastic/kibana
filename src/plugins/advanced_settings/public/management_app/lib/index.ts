/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

export { isDefaultValue } from './is_default_value';
export { toEditableConfig } from './to_editable_config';
export { getCategoryName } from './get_category_name';
export { DEFAULT_CATEGORY } from './default_category';
export { getAriaName } from './get_aria_name';
