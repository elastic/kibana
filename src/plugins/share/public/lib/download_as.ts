/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * and the Server Side Public License, v 1; you may not use this file except in
 * compliance with, at your election, the Elastic License or the Server Side
 * Public License, v 1.
 */

// @ts-ignore
import { saveAs } from '@elastic/filesaver';
import pMap from 'p-map';

export type DownloadableContent = { content: string; type: string } | Blob;

/**
 * Convenient method to use for a single file download
 * **Note**: for multiple files use the downloadMultipleAs method, do not iterate with this method here
 * @param filename full name of the file
 * @param payload either a Blob content, or a Record with a stringified content and type
 *
 * @returns a Promise that resolves when the download has been correctly started
 */
export function downloadFileAs(filename: string, payload: DownloadableContent) {
  return downloadMultipleAs({ [filename]: payload });
}

/**
 * Multiple files download method
 * @param files a Record containing one entry per file: the key entry should be the filename
 * and the value either a Blob content, or a Record with a stringified content and type
 *
 * @returns a Promise that resolves when all the downloads have been correctly started
 */
export async function downloadMultipleAs(files: Record<string, DownloadableContent>) {
  const filenames = Object.keys(files);
  const downloadQueue = filenames.map((filename, i) => {
    const payload = files[filename];
    const blob =
      // probably this is enough? It does not support Node or custom implementations
      payload instanceof Blob ? payload : new Blob([payload.content], { type: payload.type });

    // TODO: remove this workaround for multiple files when fixed (in filesaver?)
    return () => Promise.resolve().then(() => saveAs(blob, filename));
  });

  // There's a bug in some browser with multiple files downloaded at once
  // * sometimes only the first/last content is downloaded multiple times
  // * sometimes only the first/last filename is used multiple times
  await pMap(downloadQueue, (downloadFn) => Promise.all([downloadFn(), wait(50)]), {
    concurrency: 1,
  });
}
// Probably there's already another one around?
function wait(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
