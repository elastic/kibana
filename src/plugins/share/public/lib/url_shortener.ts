/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import url from 'url';
import { HttpStart } from 'kibana/public';
import { CREATE_PATH, getGotoPath } from '../../common/short_url_routes';

export async function shortenUrl(
  absoluteUrl: string,
  { basePath, post }: { basePath: string; post: HttpStart['post'] }
) {
  const parsedUrl = url.parse(absoluteUrl);
  if (!parsedUrl || !parsedUrl.path) {
    return;
  }
  const path = parsedUrl.path.replace(basePath, '');
  const hash = parsedUrl.hash ? parsedUrl.hash : '';
  const relativeUrl = path + hash;

  const body = JSON.stringify({ url: relativeUrl });

  const resp = await post(CREATE_PATH, { body });
  return url.format({
    protocol: parsedUrl.protocol,
    host: parsedUrl.host,
    pathname: `${basePath}${getGotoPath(resp.urlId)}`,
  });
}
