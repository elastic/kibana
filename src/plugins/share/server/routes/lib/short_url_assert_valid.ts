/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { parse } from 'url';
import { trim } from 'lodash';
import Boom from '@hapi/boom';

export function shortUrlAssertValid(url: string) {
  const { protocol, hostname, pathname } = parse(
    url,
    false /* parseQueryString */,
    true /* slashesDenoteHost */
  );

  if (protocol !== null) {
    throw Boom.notAcceptable(`Short url targets cannot have a protocol, found "${protocol}"`);
  }

  if (hostname !== null) {
    throw Boom.notAcceptable(`Short url targets cannot have a hostname, found "${hostname}"`);
  }

  const pathnameParts = trim(pathname === null ? undefined : pathname, '/').split('/');
  if (pathnameParts.length !== 2 || pathnameParts[0] !== 'app' || !pathnameParts[1]) {
    throw Boom.notAcceptable(
      `Short url target path must be in the format "/app/{{appId}}", found "${pathname}"`
    );
  }
}
