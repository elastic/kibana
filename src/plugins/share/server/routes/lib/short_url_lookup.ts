/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import crypto from 'crypto';
import { get } from 'lodash';

import { Logger, SavedObject, SavedObjectsClientContract } from 'kibana/server';

export interface ShortUrlLookupService {
  generateUrlId(url: string, deps: { savedObjects: SavedObjectsClientContract }): Promise<string>;
  getUrl(url: string, deps: { savedObjects: SavedObjectsClientContract }): Promise<string>;
}

export function shortUrlLookupProvider({ logger }: { logger: Logger }): ShortUrlLookupService {
  async function updateMetadata(
    doc: SavedObject,
    { savedObjects }: { savedObjects: SavedObjectsClientContract }
  ) {
    try {
      await savedObjects.update('url', doc.id, {
        accessDate: new Date().valueOf(),
        accessCount: get(doc, 'attributes.accessCount', 0) + 1,
      });
    } catch (error) {
      logger.warn('Warning: Error updating url metadata');
      logger.warn(error);
      // swallow errors. It isn't critical if there is no update.
    }
  }

  return {
    async generateUrlId(url, { savedObjects }) {
      const id = crypto
        .createHash('md5')
        .update(url)
        .digest('hex');
      const { isConflictError } = savedObjects.errors;

      try {
        const doc = await savedObjects.create(
          'url',
          {
            url,
            accessCount: 0,
            createDate: new Date().valueOf(),
            accessDate: new Date().valueOf(),
          },
          { id }
        );

        return doc.id;
      } catch (error) {
        if (isConflictError(error)) {
          return id;
        }

        throw error;
      }
    },

    async getUrl(id, { savedObjects }) {
      const doc = await savedObjects.get('url', id);
      updateMetadata(doc, { savedObjects });

      return doc.attributes.url;
    },
  };
}
