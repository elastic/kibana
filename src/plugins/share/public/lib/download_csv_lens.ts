/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
import { exporters } from '@kbn/data-plugin/public';
import type { TableInspectorAdapter } from '@kbn/lens-plugin/public/editor_frame_service/types';
import type { FormatFactory } from '@kbn/field-formats-plugin/common';
import type { IUiSettingsClient } from '@kbn/core-ui-settings-browser';
import { downloadMultipleAs } from './download_as';

declare global {
  interface Window {
    /**
     * Debug setting to test CSV download
     */
    ELASTIC_LENS_CSV_DOWNLOAD_DEBUG?: boolean;
    ELASTIC_LENS_CSV_CONTENT?: Record<string, { content: string; type: string }>;
  }
}

export async function downloadCSVs({
  activeData,
  title,
  formatFactory,
  uiSettings,
  columnsSorting,
}: {
  title: string;
  activeData: TableInspectorAdapter;
  formatFactory: FormatFactory;
  uiSettings: IUiSettingsClient;
  columnsSorting?: string[];
}) {
  if (!activeData) {
    if (window.ELASTIC_LENS_CSV_DOWNLOAD_DEBUG) {
      window.ELASTIC_LENS_CSV_CONTENT = undefined;
    }
    return;
  }
  const datatables = Object.values(activeData);
  const content = datatables.reduce<Record<string, { content: string; type: string }>>(
    (memo, datatable, i) => {
      // skip empty datatables
      if (datatable) {
        const postFix = datatables.length > 1 ? `-${i + 1}` : '';

        memo[`${title}${postFix}.csv`] = {
          content: exporters.datatableToCSV(datatable, {
            csvSeparator: uiSettings.get('csv:separator', ','),
            quoteValues: uiSettings.get('csv:quoteValues', true),
            formatFactory,
            escapeFormulaValues: false,
            columnsSorting,
          }),
          type: exporters.CSV_MIME_TYPE,
        };
      }
      return memo;
    },
    {}
  );
  if (window.ELASTIC_LENS_CSV_DOWNLOAD_DEBUG) {
    window.ELASTIC_LENS_CSV_CONTENT = content;
  }
  if (content) {
    downloadMultipleAs(content);
  }
}
