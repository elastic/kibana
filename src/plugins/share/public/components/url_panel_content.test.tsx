/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { EuiCopy, EuiRadioGroup, EuiSwitch, EuiSwitchEvent } from '@elastic/eui';

jest.mock('../lib/url_shortener', () => ({ shortenUrl: jest.fn() }));

import React from 'react';
import { shallow } from 'enzyme';

import { ExportUrlAsType, UrlPanelContent } from './url_panel_content';
import { act } from 'react-dom/test-utils';
import { shortenUrl } from '../lib/url_shortener';

const defaultProps = {
  allowShortUrl: true,
  objectType: 'dashboard',
  basePath: '',
  post: () => Promise.resolve({} as any),
};

describe('share url panel content', () => {
  test('render', () => {
    const component = shallow(<UrlPanelContent {...defaultProps} />);
    expect(component).toMatchSnapshot();
  });

  test('should enable saved object export option when objectId is provided', () => {
    const component = shallow(<UrlPanelContent {...defaultProps} objectId="id1" />);
    expect(component).toMatchSnapshot();
  });

  test('should hide short url section when allowShortUrl is false', () => {
    const component = shallow(
      <UrlPanelContent {...defaultProps} allowShortUrl={false} objectId="id1" />
    );
    expect(component).toMatchSnapshot();
  });

  test('should remove _a query parameter in saved object mode', () => {
    const component = shallow(
      <UrlPanelContent
        {...defaultProps}
        shareableUrl="http://localhost:5601/app/myapp#/?_g=()&_a=()"
        allowShortUrl={false}
        objectId="id1"
      />
    );
    act(() => {
      component.find(EuiRadioGroup).prop('onChange')!(ExportUrlAsType.EXPORT_URL_AS_SAVED_OBJECT);
    });
    expect(component.find(EuiCopy).prop('textToCopy')).toEqual(
      'http://localhost:5601/app/myapp#/?_g=()'
    );
  });

  describe('short url', () => {
    test('should generate short url and put it in copy button', async () => {
      const shortenUrlMock = shortenUrl as jest.Mock;
      shortenUrlMock.mockReset();
      shortenUrlMock.mockResolvedValue('http://localhost/short/url');

      const component = shallow(
        <UrlPanelContent
          {...defaultProps}
          shareableUrl="http://localhost:5601/app/myapp#/?_g=()&_a=()"
          objectId="id1"
        />
      );
      await act(async () => {
        component.find(EuiSwitch).prop('onChange')!(({
          target: { checked: true },
        } as unknown) as EuiSwitchEvent);
      });
      expect(shortenUrlMock).toHaveBeenCalledWith(
        'http://localhost:5601/app/myapp#/?_g=()&_a=()',
        expect.anything()
      );
      expect(component.find(EuiCopy).prop('textToCopy')).toContain('http://localhost/short/url');
    });

    test('should hide short url for saved object mode', async () => {
      const component = shallow(
        <UrlPanelContent
          {...defaultProps}
          shareableUrl="http://localhost:5601/app/myapp#/"
          objectId="id1"
        />
      );
      act(() => {
        component.find(EuiRadioGroup).prop('onChange')!(ExportUrlAsType.EXPORT_URL_AS_SAVED_OBJECT);
      });
      expect(component.exists(EuiSwitch)).toEqual(false);
    });
  });

  describe('embedded', () => {
    const asIframe = (url: string) => `<iframe src="${url}" height="600" width="800"></iframe>`;

    test('should add embedded flag to target code in snapshot mode', () => {
      const component = shallow(
        <UrlPanelContent
          {...defaultProps}
          shareableUrl="http://localhost:5601/app/myapp#/"
          isEmbedded
          allowShortUrl={false}
          objectId="id1"
        />
      );
      expect(component.find(EuiCopy).prop('textToCopy')).toEqual(
        asIframe('http://localhost:5601/app/myapp#/?embed=true')
      );
    });

    test('should add embedded flag to target code in snapshot mode with existing query parameters', () => {
      const component = shallow(
        <UrlPanelContent
          {...defaultProps}
          shareableUrl="http://localhost:5601/app/myapp#/?_g=()&_a=()"
          isEmbedded
          allowShortUrl={false}
          objectId="id1"
        />
      );
      expect(component.find(EuiCopy).prop('textToCopy')).toEqual(
        asIframe('http://localhost:5601/app/myapp#/?embed=true&_g=()&_a=()')
      );
    });

    test('should remove _a query parameter and add embedded flag in saved object mode', () => {
      const component = shallow(
        <UrlPanelContent
          {...defaultProps}
          shareableUrl="http://localhost:5601/app/myapp#/?_g=()&_a=()"
          isEmbedded
          allowShortUrl={false}
          objectId="id1"
        />
      );
      act(() => {
        component.find(EuiRadioGroup).prop('onChange')!(ExportUrlAsType.EXPORT_URL_AS_SAVED_OBJECT);
      });
      expect(component.find(EuiCopy).prop('textToCopy')).toEqual(
        asIframe('http://localhost:5601/app/myapp#/?embed=true&_g=()')
      );
    });

    test('should generate short url with embed flag and put it in copy button', async () => {
      const shortenUrlMock = shortenUrl as jest.Mock;
      shortenUrlMock.mockReset();
      shortenUrlMock.mockResolvedValue('http://localhost/short/url');

      const component = shallow(
        <UrlPanelContent
          {...defaultProps}
          isEmbedded
          shareableUrl="http://localhost:5601/app/myapp#/?_g=()&_a=()"
          objectId="id1"
        />
      );
      await act(async () => {
        component.find(EuiSwitch).prop('onChange')!(({
          target: { checked: true },
        } as unknown) as EuiSwitchEvent);
      });
      expect(shortenUrlMock).toHaveBeenCalledWith(
        'http://localhost:5601/app/myapp#/?embed=true&_g=()&_a=()',
        expect.anything()
      );
      expect(component.find(EuiCopy).prop('textToCopy')).toContain('http://localhost/short/url');
    });

    test('should hide short url for saved object mode', async () => {
      const component = shallow(
        <UrlPanelContent
          {...defaultProps}
          isEmbedded
          shareableUrl="http://localhost:5601/app/myapp#/"
          objectId="id1"
        />
      );
      act(() => {
        component.find(EuiRadioGroup).prop('onChange')!(ExportUrlAsType.EXPORT_URL_AS_SAVED_OBJECT);
      });
      expect(component.exists(EuiSwitch)).toEqual(false);
    });
  });
});

test('should show url param extensions', () => {
  const TestExtension = () => <div data-test-subj="testExtension" />;
  const extensions = [{ paramName: 'testExtension', component: TestExtension }];
  const component = shallow(
    <UrlPanelContent {...defaultProps} urlParamExtensions={extensions} objectId="id1" />
  );
  expect(component.find('TestExtension').length).toBe(1);
  expect(component).toMatchSnapshot();
});
