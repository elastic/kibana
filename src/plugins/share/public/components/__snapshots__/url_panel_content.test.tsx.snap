// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`share url panel content render 1`] = `
<I18nProvider>
  <EuiForm
    className="kbnShareContextMenu__finalPanel"
    data-test-subj="shareUrlForm"
  >
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      helpText={
        <FormattedMessage
          defaultMessage="Can't share as saved object until the {objectType} has been saved."
          id="share.urlPanel.canNotShareAsSavedObjectHelpText"
          values={
            Object {
              "objectType": "dashboard",
            }
          }
        />
      }
      label={
        <FormattedMessage
          defaultMessage="Generate the link as"
          id="share.urlPanel.generateLinkAsLabel"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiRadioGroup
        idSelected="snapshot"
        onChange={[Function]}
        options={
          Array [
            Object {
              "data-test-subj": "exportAsSnapshot",
              "id": "snapshot",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Snapshot"
                    id="share.urlPanel.snapshotLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="Snapshot URLs encode the current state of the {objectType} in the URL itself. Edits to the saved {objectType} won't be visible via this URL."
                        id="share.urlPanel.snapshotDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
            Object {
              "data-test-subj": "exportAsSavedObject",
              "disabled": true,
              "id": "savedObject",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Saved object"
                    id="share.urlPanel.savedObjectLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="You can share this URL with people to let them load the most recent saved version of this {objectType}."
                        id="share.urlPanel.savedObjectDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
          ]
        }
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label={
        <FormattedMessage
          defaultMessage="URL"
          id="share.urlPanel.urlGroupTitle"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiSpacer
        size="s"
      />
      <EuiFormRow
        data-test-subj="createShortUrl"
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        labelType="label"
      >
        <EuiFlexGroup
          gutterSize="none"
          responsive={false}
        >
          <EuiFlexItem
            grow={false}
          >
            <EuiSwitch
              checked={false}
              data-test-subj="useShortUrl"
              label={
                <FormattedMessage
                  defaultMessage="Short URL"
                  id="share.urlPanel.shortUrlLabel"
                  values={Object {}}
                />
              }
              onChange={[Function]}
            />
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <EuiIconTip
              content={
                <FormattedMessage
                  defaultMessage="We recommend sharing shortened snapshot URLs for maximum compatibility. Internet Explorer has URL length restrictions, and some wiki and markup parsers don't do well with the full-length version of the snapshot URL, but the short URL should work great."
                  id="share.urlPanel.shortUrlHelpText"
                  values={Object {}}
                />
              }
              position="bottom"
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiFormRow>
    </EuiFormRow>
    <EuiSpacer
      size="m"
    />
    <EuiCopy
      afterMessage="Copied"
      anchorClassName="eui-displayBlock"
      textToCopy="http://localhost/"
    >
      <Component />
    </EuiCopy>
  </EuiForm>
</I18nProvider>
`;

exports[`share url panel content should enable saved object export option when objectId is provided 1`] = `
<I18nProvider>
  <EuiForm
    className="kbnShareContextMenu__finalPanel"
    data-test-subj="shareUrlForm"
  >
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label={
        <FormattedMessage
          defaultMessage="Generate the link as"
          id="share.urlPanel.generateLinkAsLabel"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiRadioGroup
        idSelected="snapshot"
        onChange={[Function]}
        options={
          Array [
            Object {
              "data-test-subj": "exportAsSnapshot",
              "id": "snapshot",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Snapshot"
                    id="share.urlPanel.snapshotLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="Snapshot URLs encode the current state of the {objectType} in the URL itself. Edits to the saved {objectType} won't be visible via this URL."
                        id="share.urlPanel.snapshotDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
            Object {
              "data-test-subj": "exportAsSavedObject",
              "disabled": false,
              "id": "savedObject",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Saved object"
                    id="share.urlPanel.savedObjectLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="You can share this URL with people to let them load the most recent saved version of this {objectType}."
                        id="share.urlPanel.savedObjectDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
          ]
        }
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label={
        <FormattedMessage
          defaultMessage="URL"
          id="share.urlPanel.urlGroupTitle"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiSpacer
        size="s"
      />
      <EuiFormRow
        data-test-subj="createShortUrl"
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        labelType="label"
      >
        <EuiFlexGroup
          gutterSize="none"
          responsive={false}
        >
          <EuiFlexItem
            grow={false}
          >
            <EuiSwitch
              checked={false}
              data-test-subj="useShortUrl"
              label={
                <FormattedMessage
                  defaultMessage="Short URL"
                  id="share.urlPanel.shortUrlLabel"
                  values={Object {}}
                />
              }
              onChange={[Function]}
            />
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <EuiIconTip
              content={
                <FormattedMessage
                  defaultMessage="We recommend sharing shortened snapshot URLs for maximum compatibility. Internet Explorer has URL length restrictions, and some wiki and markup parsers don't do well with the full-length version of the snapshot URL, but the short URL should work great."
                  id="share.urlPanel.shortUrlHelpText"
                  values={Object {}}
                />
              }
              position="bottom"
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiFormRow>
    </EuiFormRow>
    <EuiSpacer
      size="m"
    />
    <EuiCopy
      afterMessage="Copied"
      anchorClassName="eui-displayBlock"
      textToCopy="http://localhost/"
    >
      <Component />
    </EuiCopy>
  </EuiForm>
</I18nProvider>
`;

exports[`share url panel content should hide short url section when allowShortUrl is false 1`] = `
<I18nProvider>
  <EuiForm
    className="kbnShareContextMenu__finalPanel"
    data-test-subj="shareUrlForm"
  >
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label={
        <FormattedMessage
          defaultMessage="Generate the link as"
          id="share.urlPanel.generateLinkAsLabel"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiRadioGroup
        idSelected="snapshot"
        onChange={[Function]}
        options={
          Array [
            Object {
              "data-test-subj": "exportAsSnapshot",
              "id": "snapshot",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Snapshot"
                    id="share.urlPanel.snapshotLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="Snapshot URLs encode the current state of the {objectType} in the URL itself. Edits to the saved {objectType} won't be visible via this URL."
                        id="share.urlPanel.snapshotDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
            Object {
              "data-test-subj": "exportAsSavedObject",
              "disabled": false,
              "id": "savedObject",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Saved object"
                    id="share.urlPanel.savedObjectLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="You can share this URL with people to let them load the most recent saved version of this {objectType}."
                        id="share.urlPanel.savedObjectDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
          ]
        }
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label={
        <FormattedMessage
          defaultMessage="URL"
          id="share.urlPanel.urlGroupTitle"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiSpacer
        size="s"
      />
    </EuiFormRow>
    <EuiSpacer
      size="m"
    />
    <EuiCopy
      afterMessage="Copied"
      anchorClassName="eui-displayBlock"
      textToCopy="http://localhost/"
    >
      <Component />
    </EuiCopy>
  </EuiForm>
</I18nProvider>
`;

exports[`should show url param extensions 1`] = `
<I18nProvider>
  <EuiForm
    className="kbnShareContextMenu__finalPanel"
    data-test-subj="shareUrlForm"
  >
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label={
        <FormattedMessage
          defaultMessage="Generate the link as"
          id="share.urlPanel.generateLinkAsLabel"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiRadioGroup
        idSelected="snapshot"
        onChange={[Function]}
        options={
          Array [
            Object {
              "data-test-subj": "exportAsSnapshot",
              "id": "snapshot",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Snapshot"
                    id="share.urlPanel.snapshotLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="Snapshot URLs encode the current state of the {objectType} in the URL itself. Edits to the saved {objectType} won't be visible via this URL."
                        id="share.urlPanel.snapshotDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
            Object {
              "data-test-subj": "exportAsSavedObject",
              "disabled": false,
              "id": "savedObject",
              "label": <EuiFlexGroup
                gutterSize="none"
                responsive={false}
              >
                <EuiFlexItem
                  grow={false}
                >
                  <FormattedMessage
                    defaultMessage="Saved object"
                    id="share.urlPanel.savedObjectLabel"
                    values={Object {}}
                  />
                </EuiFlexItem>
                <EuiFlexItem
                  grow={false}
                >
                  <EuiIconTip
                    content={
                      <FormattedMessage
                        defaultMessage="You can share this URL with people to let them load the most recent saved version of this {objectType}."
                        id="share.urlPanel.savedObjectDescription"
                        values={
                          Object {
                            "objectType": "dashboard",
                          }
                        }
                      />
                    }
                    position="bottom"
                  />
                </EuiFlexItem>
              </EuiFlexGroup>,
            },
          ]
        }
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      key="testExtension"
      labelType="label"
    >
      <TestExtension
        setParamValue={[Function]}
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label={
        <FormattedMessage
          defaultMessage="URL"
          id="share.urlPanel.urlGroupTitle"
          values={Object {}}
        />
      }
      labelType="label"
    >
      <EuiSpacer
        size="s"
      />
      <EuiFormRow
        data-test-subj="createShortUrl"
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        labelType="label"
      >
        <EuiFlexGroup
          gutterSize="none"
          responsive={false}
        >
          <EuiFlexItem
            grow={false}
          >
            <EuiSwitch
              checked={false}
              data-test-subj="useShortUrl"
              label={
                <FormattedMessage
                  defaultMessage="Short URL"
                  id="share.urlPanel.shortUrlLabel"
                  values={Object {}}
                />
              }
              onChange={[Function]}
            />
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <EuiIconTip
              content={
                <FormattedMessage
                  defaultMessage="We recommend sharing shortened snapshot URLs for maximum compatibility. Internet Explorer has URL length restrictions, and some wiki and markup parsers don't do well with the full-length version of the snapshot URL, but the short URL should work great."
                  id="share.urlPanel.shortUrlHelpText"
                  values={Object {}}
                />
              }
              position="bottom"
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiFormRow>
    </EuiFormRow>
    <EuiSpacer
      size="m"
    />
    <EuiCopy
      afterMessage="Copied"
      anchorClassName="eui-displayBlock"
      textToCopy="http://localhost/"
    >
      <Component />
    </EuiCopy>
  </EuiForm>
</I18nProvider>
`;
